### PR TITLE
77650 get functional roles for ipo

### PR DESCRIPTION
**Pull request by llykkelarsen**
move mainapi project into common folder for non-ipo apis from which we get data ('foreignapi')
create endpoint to get functional roles from library api based on classification

### CODE REVIEW

**Review by bratli — APPROVED** [review]

**Comment by bratli** [comment]
Fix failing tests

File: src/Equinor.ProCoSys.IPO.WebApi/Authentication/Authenticator.cs [hunk 1]
```diff
@@ -40,14 +41,38 @@ public async ValueTask<string> GetBearerTokenOnBehalfOfCurrentUserAsync()
...
+            return _requestToken;
+        }
+
+        public async ValueTask<string> GetBearerTokenForLibraryApiOnBehalfOfCurrentUserAsync()
+        {
+            if (_canUseOnBehalfOf)
             {
-                return _requestToken;
+                if (_onBehalfOfUserTokenForLibraryApi == null)
+                {
```
Comment (paal123) [inline src/Equinor.ProCoSys.IPO.WebApi/Authentication/Authenticator.cs:57]: Extract as a method injecting scope, as that is the only difference for mainapi and libraryapi scope

File: src/Equinor.ProCoSys.IPO.WebApi/Controllers/Participants/ParticipantsController.cs [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+        /// </summary>
+        /// <param name="plant"></param>
+        /// <returns>All ProCoSys functional roles which have classification IPO</returns>
+        [Authorize(Roles = Permissions.LIBRARY_FUNCTIONAL_ROLE_READ)]
+        [HttpGet("/FunctionalRoles")]
+        public async Task<ActionResult<List<ProCoSysFunctionalRoleDto>>> GetFunctionalRolesForIpo(
+            [FromHeader(Name = CurrentPlantMiddleware.PlantHeader)] [Required]
+            string plant)
+        {
+            const string classification = "IPO";
```
Comment (paal123) [inline src/Equinor.ProCoSys.IPO.WebApi/Controllers/Participants/ParticipantsController.cs:32]: Move out as member variable.

File: src/Equinor.ProCoSys.IPO.WebApi/Permissions.cs [hunk 3]
```diff
@@ -5,5 +5,6 @@ public class Permissions
         public const string COMMPKG_READ = "COMMPKG/READ";
         public const string PROJECT_READ = "PROJECT/READ";
         public const string MCPKG_READ = "MCPKG/READ";
+        public const string LIBRARY_FUNCTIONAL_ROLE_READ = "LIBRARY_FUNCTIONAL_ROLE/READ";
```
Comment (paal123) [inline src/Equinor.ProCoSys.IPO.WebApi/Permissions.cs:8]: Could we get away with just using LIBRARY/READ? for reading library data.

**Review by paal123 — APPROVED** [review]
With suggestions

File: src/Equinor.ProCoSys.IPO.WebApi/Permissions.cs (hunk 3 above)
Comment (llykkelarsen) [reply src/Equinor.ProCoSys.IPO.WebApi/Permissions.cs:8]: I don't think so as there's no permission called just 'LIBRARY'

**Pull request merged**